packages/webpack-cli/src/plugins/dotenv-webpack-plugin.ts

```diff
@@ -0,0 +1,253 @@
+    this.#logger = compiler.getInfrastructureLogger("dotenv-webpack-plugin");
+    if (!this.#options.paths) {
+      this.#options.paths = [
+        ".env",
+        ".env.local",
+        `.env.${compiler.options.mode}`,
+        `.env.${compiler.options.mode}.local`,
+      ];
+    }
+    compiler.hooks.thisCompilation.tap("dotenv-webpack-plugin", (compilation) => {
+      const setupAsyncOperations = async () => {
+        const cache = compilation.getCache("dotenv-webpack-plugin-compiler");
+        if (await cache.getPromise("dotenv-webpack-plugin-data", null)) {
+          new DefinePlugin(await cache.getPromise("dotenv-webpack-plugin-data", null)).apply(
+            compiler,
+          );
+        } else {
+          const variables = await this.#gatherVariables(compilation);
+          const target: string =
+            typeof compiler.options.target == "string" ? compiler.options.target : "";
+          const data = this.#formatData({
+            variables,
+            target: target,
+          });
+          new DefinePlugin(data).apply(compiler);
+          await cache.storePromise("dotenv-webpack-plugin-data", null, data);
+        }
+      };
+      setupAsyncOperations().catch((err) => {
+        this.#logger.error(`Error in dotenv-webpack-plugin: ${err}`);
+      });
+    });
+  }
+  async #gatherVariables(compilation: any): Promise<EnvVariables> {
+    const { allowEmptyValues, safe } = this.#options;
+    const vars: EnvVariables = this.#initializeVars();
+
+    const { env, blueprint } = await this.#getEnvs(compilation);
+
+    Object.keys(blueprint).forEach((key) => {
+      const value = Object.prototype.hasOwnProperty.call(vars, key) ? vars[key] : env[key];
+
+      if (
+        (typeof value === "undefined" || value === null || (!allowEmptyValues && value === "")) &&
+        safe
+      ) {
+        compilation.errors.push(new Error(`Missing environment variable: ${key}`));
+      } else {
+        vars[key] = value;
+      }
+      if (safe) {
+        Object.keys(env).forEach((key) => {
+          if (!Object.prototype.hasOwnProperty.call(vars, key)) {
+            vars[key] = env[key];
+          }
+        });
+      }
+    });
+
+    return vars;
+  }
+
+  #initializeVars(): EnvVariables {
+    return this.#options.systemvars
+      ? Object.fromEntries(Object.entries(process.env).map(([key, value]) => [key, value ?? ""]))
+      : {};
+  }
+
+  merge(apply = {}, defaults = {}) {
+    Object.assign({}, defaults, apply);
+  }
+
+  parse(src: string, defaultSrc = "") {
+    const parsedSrc = dotenv.parse(src);
+    const parsedDefault = dotenv.parse(defaultSrc);
+    const parsed = {};
+    Object.assign(parsed, parsedDefault, parsedSrc);
+    return parsed;
+  }
+
+  async #getEnvs(compilation: any): Promise<{ env: EnvVariables; blueprint: EnvVariables }> {
+    const { paths, safe } = this.#options;
+
+    const env: EnvVariables = {};
+    let blueprint: EnvVariables = {};
+    if (paths) {
```

Comment:
Avoid checks in your code, make all validation in constructor, it decreases readability of your code